src/Microsoft.VisualStudio.Threading/AsyncBarrier.cs

```diff
@@ -55,23 +62,31 @@ public Task SignalAndWait()
                 // Unleash everyone that preceded this one.
                 while (this.waiters.Count > 0)
                 {
-                    Task.Factory.StartNew(
-                        state => ((TaskCompletionSource<EmptyStruct>)state!).SetResult(default(EmptyStruct)),
-                        this.waiters.Pop(),
-                        CancellationToken.None,
-                        TaskCreationOptions.None,
-                        TaskScheduler.Default);
+                    this.waiters.Pop().TrySetResult(default);
                 }
 
                 // And allow this one to continue immediately.
-                return Task.CompletedTask;
+                return new ValueTask(cancellationToken.IsCancellationRequested
+                    ? Task.FromCanceled(cancellationToken)
+                    : Task.CompletedTask);
             }
             else
             {
                 // We need more folks. So suspend this caller.
-                var tcs = new TaskCompletionSource<EmptyStruct>();
+                TaskCompletionSource<EmptyStruct> tcs = new(TaskCreationOptions.RunContinuationsAsynchronously);
+#if NET
+                cancellationToken.Register(static (tcs, ct) => ((TaskCompletionSource<EmptyStruct>)tcs!).TrySetCanceled(ct), tcs);
```

Comment:
💡 Another option is to just delegate this to `WithCancellation`

Reply:
True, although that would create yet another `TaskCompletionSource<T>`, so doing it manually is slightly more efficient.